From oliver@neukum.org  Fri Jul 10 15:17:50 2009
From: Oliver Neukum <oliver@neukum.org>
Date: Thu, 2 Jul 2009 12:07:07 +0200
Subject: USB: legousbtower: make poll notice disconnect
To: greg@kroah.com, USB list <linux-usb@vger.kernel.org>, starblue@users.sourceforge.net, davidgsf@sourceforge.net
Message-ID: <200907021207.07770.oliver@neukum.org>
Content-Disposition: inline


poll needs to return an error if a device is disconnected
  - make poll check for device's presence
  - wake all waiters in disconnect

Signed-off-by: Oliver Neukum <oliver@neukum.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/misc/legousbtower.c |    6 ++++++
 1 file changed, 6 insertions(+)

--- a/drivers/usb/misc/legousbtower.c
+++ b/drivers/usb/misc/legousbtower.c
@@ -552,6 +552,9 @@ static unsigned int tower_poll (struct f
 
 	dev = file->private_data;
 
+	if (!dev->udev)
+		return POLLERR | POLLHUP;
+
 	poll_wait(file, &dev->read_wait, wait);
 	poll_wait(file, &dev->write_wait, wait);
 
@@ -1025,6 +1028,9 @@ static void tower_disconnect (struct usb
 		tower_delete (dev);
 	} else {
 		dev->udev = NULL;
+		/* wake up pollers */
+		wake_up_interruptible_all(&dev->read_wait);
+		wake_up_interruptible_all(&dev->write_wait);
 		mutex_unlock(&dev->lock);
 	}
 
